docs/basics/part-6-using-docker-with-earthly.md

```diff
@@ -0,0 +1,193 @@
+
+### Pulling an Image
+```Dockerfile
+hello:
+    FROM earthly/dind:alpine
+    WITH DOCKER --pull hello-world
+        RUN docker run hello-world
+    END
+
+```
+You can see in the command above that we can pass a flag to `WITH DOCKER` telling it to pull an image from Docker Hub. We can pass other flags to load in artifacts built by other targets `--load` or even images defined by docker-compose `--compose`. These images will be available within the context of `WITH DOCKER`'s docker daemon.
+
+### Loading an Image
+We can load in an image created by another target with the `--load` flag.
+
+```Dockerfile
+my-hello-world:
+    FROM ubuntu
+    CMD echo 'hello world'
+    SAVE IMAGE my-hello:latest
+
+hello:
+    FROM earthly/dind:alpine
+    WITH DOCKER --load hello:latest=+my-hello-world
+        RUN docker run hello
+    END
+```
+
+## A Real World Example
+
+One common use case for `WITH DOCKER` is running integration tests that require other services. In this case we need to set up a redis service for our tests.
+
+### Using Docker Compose
+
+`docker-compose.yml`
+```yml
+version: "3"
+services:
+  redis:
+    container_name: local-redis
+    image: redis:6.0-alpine
+    ports:
+      - 127.0.0.1:6379:6379
+    hostname: redis
+    healthcheck:
+      test: ["CMD", "curl", "-f", "http://localhost:6379"]
+      interval: 1s
+      timeout: 10s
+      retries: 5
+
+```
+`main.go`
+```go
+package main
+
+import (
+	"github.com/sirupsen/logrus"
+)
+
+var howCoolIsEarthly = "IceCool"
+
+func main() {
+	logrus.Info("hello world")
+}
+```
+`main_integration_test.go`
+```go
+package main
+
+import (
+	"context"
+	"testing"
+
+	"github.com/go-redis/redis/v8"
+	"github.com/stretchr/testify/require"
+)
+
+func TestIntegration(t *testing.T) {
+	ctx := context.Background()
+	rdb := redis.NewClient(&redis.Options{
+		Addr:     "localhost:6379",
+		Password: "", // no password set
+		DB:       0,  // use default DB
+	})
+
+	err := rdb.Set(ctx, "howCoolIsEarthly", howCoolIsEarthly, 0).Err()
+	if err != nil {
+		panic(err)
+	}
+
+	resultFromDB, err := rdb.Get(ctx, "howCoolIsEarthly").Result()
+	if err != nil {
+		panic(err)
+	}
+	require.Equal(t, howCoolIsEarthly, resultFromDB)
+}
+```
+
+```Dockerfile
+VERSION 0.6
+FROM golang:1.15-alpine3.13
+WORKDIR /go-example
+
+deps:
+    COPY go.mod go.sum ./
+    RUN go mod download
+    SAVE ARTIFACT go.mod AS LOCAL go.mod
+    SAVE ARTIFACT go.sum AS LOCAL go.sum
+
+build:
+    FROM +deps
+    COPY main.go .
+    RUN go build -o build/go-example main.go
+    SAVE ARTIFACT build/go-example /go-example AS LOCAL build/go-example
+
+integration-test:
+    FROM +deps
+    COPY main.go .
+    COPY main_integration_test.go .
+    COPY docker-compose.yml ./
+    WITH DOCKER --compose docker-compose.yml
+        RUN CGO_ENABLED=0 go test github.com/earthly/earthly/examples/go
+    END
```

Comment:
For this one I ended up using --load to load a special image build to run the tests. I thought this would be a little more straight forward and also showcase the --load command. Lmk if you think this works. I can certainly write somethign up for this example if you'd rather use this one.